core/src/main/java/bio/terra/pearl/core/service/workflow/TaskConfigDispatcher.java

```diff
@@ -0,0 +1,338 @@
+import bio.terra.pearl.core.service.participant.PortalParticipantUserService;
+import bio.terra.pearl.core.service.rule.EnrolleeContext;
+import bio.terra.pearl.core.service.rule.EnrolleeContextService;
+import bio.terra.pearl.core.service.search.EnrolleeSearchContext;
+import bio.terra.pearl.core.service.search.EnrolleeSearchExpressionParser;
+import bio.terra.pearl.core.service.study.StudyEnvironmentService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.BeanUtils;
+import org.springframework.stereotype.Service;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
+
+/**
+ * Handles dispatching tasks to enrollees based on task configurations.
+ */
+@Service
+@Slf4j
+public abstract class TaskConfigDispatcher<T extends TaskConfig> {
+    private final StudyEnvironmentService studyEnvironmentService;
+    private final ParticipantTaskService participantTaskService;
+    private final EnrolleeService enrolleeService;
+    private final EnrolleeSearchExpressionParser enrolleeSearchExpressionParser;
+    private final EnrolleeContextService enrolleeContextService;
+    private final PortalParticipantUserService portalParticipantUserService;
+
+
+    public TaskConfigDispatcher(StudyEnvironmentService studyEnvironmentService,
+                                ParticipantTaskService participantTaskService,
+                                EnrolleeService enrolleeService,
+                                EnrolleeSearchExpressionParser enrolleeSearchExpressionParser,
+                                EnrolleeContextService enrolleeContextService,
+                                PortalParticipantUserService portalParticipantUserService) {
+        this.studyEnvironmentService = studyEnvironmentService;
+        this.participantTaskService = participantTaskService;
+        this.enrolleeService = enrolleeService;
+        this.enrolleeSearchExpressionParser = enrolleeSearchExpressionParser;
+        this.enrolleeContextService = enrolleeContextService;
+        this.portalParticipantUserService = portalParticipantUserService;
+    }
+
+    protected abstract List<T> findTaskConfigsByStudyEnvironment(UUID studyEnvId);
+
+    protected abstract Optional<T> findTaskConfigForAssignDto(UUID studyEnvironmentId, ParticipantTaskAssignDto participantTaskAssignDto);
+
+    protected abstract TaskType getTaskType(T taskDispatchConfig);
+
+    public void assignScheduledTasks() {
+        List<StudyEnvironment> studyEnvironments = studyEnvironmentService.findAll();
+        for (StudyEnvironment studyEnv : studyEnvironments) {
+            assignScheduledTasks(studyEnv);
+        }
+    }
+
+    public List<ParticipantTask> assign(ParticipantTaskAssignDto assignDto,
+                                        UUID studyEnvironmentId,
+                                        ResponsibleEntity operator) {
```

Comment:
Something we could do is also add a `boolean match(AssignDto)` function and in our `ParticipantTaskExtService` iterate through the `List<TaskConfigDispatcher>` provided by spring. That way assigndtos can automatically "just work" with new tasks. That could be how we do manual task assignment